forage-android/src/main/java/com/joinforage/forage/android/ui/ForagePINEditText.kt

```diff
@@ -25,48 +25,71 @@ class ForagePINEditText @JvmOverloads constructor(
     attrs: AttributeSet? = null,
     defStyleAttr: Int = R.attr.foragePanEditTextStyle
 ) : AbstractForageElement(context, attrs, defStyleAttr) {
-    private var vault: VaultWrapper
+    private val _linearLayout: LinearLayout
+    private val btVaultWrapper: BTVaultWrapper
+    private val vgsVaultWrapper: VGSVaultWrapper
+    private var _SET_ONLY_vault: VaultWrapper? = null
+    private val vault: VaultWrapper
+        get() {
+            if (_SET_ONLY_vault == null) {
+                TODO("throw an error here since that means setForageContext was not called")
+                throw Exception("need to call setForageContext first!")
+            }
+            return _SET_ONLY_vault!!
+        }
```

Comment:
I left [a comment](https://github.com/teamforage/forage-android-sdk/pull/97#discussion_r1331955281) in your other PR about the dev experience and the risk of crashing in production if we throw in cases like this. 

Can you assure that this won't be an issue in production?

Reply:
What is the `_SET_ONLY` member for? 
Instead of relying on `_SET_ONLY_vault` being set, can we make the `_config` member from the parent a protected member and check its value instead?

This cleanup work can be deferred to a separate ticket if this release is time-sensitive.

Reply:
Great question - I will definitely leave a comment to clarify. In the mean time, the relationship between _SET_ONLY_vault and vault is as follows:

The actual code in `ForagePINEditText` wants to use vault - that's
all it cares about. However, vault is the result of a decision
of whether we wish to use Basis Theory or VGS. The answer to this decision
depends on Launch Darkly. We can only get an answer from launch
darkly, once we know the environment we're in, which depends on
the forageConfig being set on this instance

Given the above, vault is an abstraction that allows the 
rest of the code in `ForagePANEditText` to work with a
certain (non-nullable) answer. It relies on the 
backing, write-only property `_SET_ONLY_vault`. `_SET_ONLY_vault`
will only be instantiated after ForageConfig was set on this instance
We enforce the abstraction that vault exists, by throwing a runtime
exception if vault, which is a computed property, was ever accessed
at a point `_SET_ONLY_vault` is null

Reply:
Regarding this:

> Instead of relying on _SET_ONLY_vault being set, can we make the _config member from the parent a protected member and check its value instead?

Let's circle back after the release to talk about this. I made a ticket here for it [here](https://linear.app/joinforage/issue/FX-588/android-consider-alternative-to-set-only-vault)

Reply:
Regarding this:

> the risk of crashing in production if we throw in cases like this.
>
> Can you assure that this won't be an issue in production?

the typical contract is that that everyone expects and is OK with is the contract
that the Android runtime enforces:

> You must initialize all of the state necessary for your class to work nominally
> within the `init { ... }` method.

As mentioned above:

> The actual code in ForagePINEditText wants to use vault - that's
> all it cares about. However, vault is the result of a decision
> of whether we wish to use Basis Theory or VGS. The answer to this decision
> depends on Launch Darkly. We can only get an answer from launch
> darkly, once we know the environment we're in, which depends on
> the forageConfig being set on this instance

This mean we need to make a new, more demanding contract than the typical one
developers expect:

> The Forage Android SDK guarantees that every ForageElement will work nominally
> if both `init { ... }` and `setForageConfig { ... }` get called prior before
> executing any other methods

This second contract, by definition of "nominal", guarantees that if merchants
call `setForageConfig { ... }` before they call anything else (recall that Android
runtime already calls `init { ... }`), no **unexpected** runtime exceptions will
be thrown.

Well that's all good in theory land but we care about in practice. The question
becomes how do we enforce the guarantee that if a merchant calls
`setForageConfig` before calling anything else, then no exceptions will be thrown?

The answer is basically enforcing conventions or Forage developers and a little
bit of type checking by the compiler:

Reducing Risk with Forage Developer Convention
`_SET_ONLY_vault` is a mutable variable and in theory any code in
`ForagePINEditText` could set it to `null` and that would break things. But, by
giving a variable such an ugly glaring name as _SET_ONLY_vault, the risk of such
an accident happening is significantly reduce (unlikely that your IDE will
suggest this variable if what you really want is to read `vault`) and the second
line of defense is that somebody reviewing a code change is likely to see an
ugly variable like `_SET_ONLY_vault` an raise a question.

Reducing Risk with Compile Time Guarantees
The `vault` variable that will be in high demand, is read-only so the
compiler enforces that it cannot be accidentally set to `null` (which would
totally lead to bugs otherwise).

We further reduce risk by not burdening downstream consumers of `vault` to have
to handle the possibility for `vault` being `null`. If every piece of code that
consumed `vault` needing to handle `null`, I agree that would create a risk that
an uncaught exception happens in `prod`. But, by throwing a runtime exception
if `_SET_ONLY_vault` is null in the `get()` definition of `vault`, the compiler
is acknowledges that there is no way any downstream code will accidentally see
`vault == null`.

You concern about the consequence of throwing unchecked exceptions at runtime is
legitimate. However, I feel our current strategy substantially reduces the risks
Let me know what you think and/or if you want to have a sync chat about this